hubblestack/transport/ipc.py

```diff
@@ -0,0 +1,372 @@
+                                case it is used as the port for a tcp
+                                localhost connection.
+    '''
+    def __init__(self, socket_path, io_loop=None):
+        '''
+        Create a new IPC client
+
+        IPC clients cannot bind to ports, but must connect to
+        existing IPC servers. Clients can then send messages
+        to the server.
+
+        '''
+        self.io_loop = io_loop or tornado.ioloop.IOLoop.current()
+        self.socket_path = socket_path
+        self._closing = False
+        self.stream = None
+        # msgpack deprecated `encoding` starting with version 0.5.2
+        if msgpack.version >= (0, 5, 2):
+            # Under Py2 we still want raw to be set to True
+            msgpack_kwargs = {'raw': False}
+        else:
+            msgpack_kwargs = {'encoding': 'utf-8'}
+        self.unpacker = msgpack.Unpacker(**msgpack_kwargs)
+
+    def connected(self):
+        return self.stream is not None and not self.stream.closed()
+
+    def connect(self, callback=None, timeout=None):
+        '''
+        Connect to the IPC socket
+        '''
+        if hasattr(self, '_connecting_future') and not self._connecting_future.done():  # pylint: disable=E0203
+            future = self._connecting_future  # pylint: disable=E0203
+        else:
+            if hasattr(self, '_connecting_future'):
+                # read previous future result to prevent the "unhandled future exception" error
+                self._connecting_future.exception()  # pylint: disable=E0203
+            future = tornado.concurrent.Future()
+            self._connecting_future = future
+            self._connect(timeout=timeout)
+
+        if callback is not None:
+            def handle_future(future):
+                response = future.result()
+                self.io_loop.add_callback(callback, response)
+            future.add_done_callback(handle_future)
+
+        return future
+
+    @tornado.gen.coroutine
+    def _connect(self, timeout=None):
+        '''
+        Connect to a running IPCServer
+        '''
+        if isinstance(self.socket_path, int):
+            sock_type = socket.AF_INET
+            sock_addr = ('127.0.0.1', self.socket_path)
+        else:
+            sock_type = socket.AF_UNIX
+            sock_addr = self.socket_path
+
+        self.stream = None
+        if timeout is not None:
+            timeout_at = time.time() + timeout
+
+        while True:
+            if self._closing:
+                break
+
+            if self.stream is None:
+                with hubblestack.utils.asynchronous.current_ioloop(self.io_loop):
+                    self.stream = IOStream(
+                        socket.socket(sock_type, socket.SOCK_STREAM)
+                    )
+            try:
+                log.trace('IPCClient: Connecting to socket: %s', self.socket_path)
+                yield self.stream.connect(sock_addr)
```

Comment:
I hope we are copying code which we are using.
It seems, this code is related to salt/minion which we don't use.